spec/controllers/products_controller_spec.rb

```diff
@@ -0,0 +1,198 @@
+require 'rails_helper'
+
+RSpec.describe ProductsController, type: :controller do
+  before :each do
+    @product = Product.create(name: "starfish", price: 3, merchant_id: 1, description: "A starfish!")
+  end
+
+  let(:create_params) do
+  { merchant_id: 1,
+    product: {
+      name: "Dogfish",
+      price: 2,
+      description: "new description"
+    }
+  }
+  end
+
+  let(:update_params) do
+  { merchant_id: 1,
+    id: 1,
+    product: {
+      name: "Catfish",
+      price: 2,
+      description: "new description"
+    }
+  }
+  end
+
+  let(:bad_params) do
+    { merchant_id: 1,
+      id: 1,
+      product: {
+        name: "",
+      }
+    }
+  end
+
```

Comment:
I'd suggest using `valid_<object_name>` and `invalid_<object_name>` rather than `good_<object_name>` and `bad_<object_name>`. Descriptive naming is *super* helpful when you're working in a larger code base... or looking at code you wrote a while ago. :wink: 

Also see my almost identical comment in the product model specs.